### PR TITLE
moderation: add cc convenience functions

### PR DESCRIPTION
**Pull request by l-zeuch**
Add convenience functions for moderation purposes.

This patchset only adds the `getWarnings` function which returns a slice of
warning (objects) the given user has.

Future patches will slowly add more such convenience functions, like
`muteUser` and similar. This is just to test the waters how well it is
received.

There is also a quite popular suggestion on the support server, see
here:
https://canary.discord.com/channels/166207328570441728/356486960417734666/1108833432953438319

----
This patchset is a combination of several commits:
    * moderation: add mod convenience functions for CCs
    * common/templates: export TargetUserID

The export of TargetUserID in common/templates is simply because I
wanted to use that function and I saw no point in duplicating code.
However, it is quite possible that I overlooked a similar function. In
which case, please do share its signature :D

Signed-off-by: Luca Zeuch <l-zeuch@email.de>

### CODE REVIEW

**Comment by mrbentarikau** [comment]
tmplextensions.go ... would be better in plural, like other such extension files already have.

**Comment by l-zeuch** [comment]
> tmplextensions.go ... would be better in plural, like other such extension files already have.

Done in 6af1a9d022 :+1:

**Comment by phenpessoa** [comment]
I know other similar funcs do not check if `TargetUserID` returned a valid ID, but I think it is worth it checking here. 

Also, is the call to `IncreaseCheckGenericAPICall` necessary? I don't think this template func makes any call to the Discord API.

**Comment by mrbentarikau** [comment]
maybe also rename cc-function to getWarnings not getWarns.

**Comment by mrbentarikau** [comment]
another thing to consider is that "moderation_warnings" table you are quering over contains links to logs, these links in that table are not deleted even after 30-day period is over and "message_logs2" and "messages2" entries get deleted, therefore these all point to non-valid URLs causing users redbars and confusion.

**Comment by l-zeuch** [comment]
> I know other similar funcs do not check if `TargetUserID` returned a
> valid ID, but I think it is worth it checking here.

`templates.TargetUserID` is a pretty dumb function in terms of testing
whether the given input is even a valid user, so checking things here
makes no real difference.
If anything, a change to the callee would be necessary to achieve a
robust behaviour in terms of "is this user ID even a valid one".

> Also, is the call to `IncreaseCheckGenericAPICall` necessary? I don't
> think this template func makes any call to the Discord API.

No, it is not necessary, I simply had it there as a reminder to perhaps
introduce a new counter for all moderation things. I'm just not yet sure
what the best limit for this would be.

> another thing to consider is that "moderation_warnings" table you are
> quering over contains links to logs, these links in that table are
> not deleted after 30-day period is over and "message_logs2" and
> "messages2" entries get deleted, therefore these all point to
> non-valid URLs causing users redbars and confusion.

FWIW, automatic log deletion should be documented behaviour for
userland. Meaning, any confusion is user error.

> for "warnings" stdcommand there's a date check-built in inside
> moderation/commands.go and -30 day links are not displayed.

I'll implement this check here anyway, shouldn't be too hard and it
doesn't hurt anyone either.

**Comment by phenpessoa** [comment]
> templates.TargetUserID is a pretty dumb function in terms of testing
> whether the given input is even a valid user, so checking things here
> makes no real difference.
> If anything, a change to the callee would be necessary to achieve a
> robust behaviour in terms of "is this user ID even a valid one".

I'm not sure I agree with `so checking things here makes no real difference`.
It does make a difference. You will be making a database call that you already knew beforehand wouldn't return anything because the ID is not valid.

**Comment by l-zeuch** [comment]
> > templates.TargetUserID is a pretty dumb function in terms of testing
> > whether the given input is even a valid user, so checking things here
> > makes no real difference.
> > If anything, a change to the callee would be necessary to achieve a
> > robust behaviour in terms of "is this user ID even a valid one".
> 
> I'm not sure I agree with `so checking things here makes no real difference`. 
> It does make a difference. You will be making a database call that you already
> knew beforehand wouldn't return anything because the ID is not valid.

Give `templates.TargetUserID` an ID that is *definitely not* valid. It will return
the value it received. If anything, the check for "is this input even a user ID" 
should be made inside that function and then return an appropriate value.

Only then can we reliably check at the callsite if we gave it a sensible value.

**Comment by phenpessoa** [comment]
> Give templates.TargetUserID an ID that is definitely not valid. It will return the value it received.

I think I'm missing something, because this doesn't seem to be the case at all.
It will not return whatever it received. It will return an int64. 
From what I can tell it will always return 0 given an invalid input.
So we just need to check if the return value is 0, if it is, do not make the DB call.

I'm not saying we should validate the user input here.
I'm just saying we should return an empty response in case we already know that is what the DB will return anyway.

**Comment by l-zeuch** [comment]
> > Give templates.TargetUserID an ID that is definitely not valid.
> > It will return the value it received.
> 
> I think I'm missing something, because this doesn't seem to be the
case at all. It will not return whatever it received. It will return an
> int64. From what I can tell it will always return 0 given an invalid
> input. So we just need to check if the return value is 0, if it is,
> do not make the DB call.
> 
> I'm not saying we should validate the user input here. I'm just
> saying we should return an empty response in case we already know
> that is what the DB will return anyway.

Looks like we've been talking about different things here. Now your
comment makes a lot more sense.

Obviously I do agree with checking for actually bogus input here.

File: moderation/tmplextensions.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	return func(target interface{}) ([]*WarningModel, error) {
+		if ctx.IncreaseCheckCallCounterPremium("cc_moderation", 5, 10) {
+			return nil, templates.ErrTooManyCalls
+		}
+
+		gID := ctx.GS.ID
+		var warns []*WarningModel
+		targetID := templates.TargetUserID(target)
+		if targetID == 0 {
+			return nil, errors.New(fmt.Sprintf("wrong type for user; expected int64, string, or discordgo.User, got %T", target))
```
[the remark this answers is not in the document]
Comment (l-zeuch) [reply moderation/tmplextensions.go]: Should be better now, I think.

**Pull request merged**